More doctest refinements.

Erotemic opened the pull request:
Follow up to #82797

Now that the doctests themselves are in a better state, we should be able to enable xdoctest on the CI so they stay that way.

@ezyang @vadimkantorov

Erotemic committed:
Enable xdoctest in CI

ezyang commented on .jenkins/pytorch/macos-test.sh:
```diff
@@ -16,9 +16,8 @@ fi
-  # TODO: enable xdoctest later
-  # xdoctest
+  pytest-rerunfailures \
+  xdoctest
```
probably better to pin this version

ezyang commented on .jenkins/pytorch/win-test-helpers/setup_pytorch_env.bat:
```diff
@@ -36,8 +36,7 @@ popd
 
-pip install "ninja==1.10.0.post1" future "hypothesis==5.35.1" "expecttest==0.1.3" "librosa>=0.6.2" "scipy==1.6.3" psutil pillow "unittest-xml-reporting<=3.2.0,>=2.0.0" pytest pytest-xdist pytest-rerunfailures 
-:: # TODO: enable xdoctest later
+pip install "ninja==1.10.0.post1" future "hypothesis==5.35.1" "expecttest==0.1.3" "librosa>=0.6.2" "scipy==1.6.3" psutil pillow "unittest-xml-reporting<=3.2.0,>=2.0.0" pytest pytest-xdist pytest-rerunfailures xdoctest 
```
ditto here (version pinning)

Erotemic replied:
I pinned the version everywhere.

Erotemic committed:
Pin xdoctest version

Erotemic commented:
Tests are passing, but I also don't see evidence that xdoctest is actually enabled and running.

I was assuming that `pytest` was used as the main runner, and all I needed to do was add the `--xdoctest` flag to enable the plugin, but is that not how the tests are run on the CI? Do I need to execute the doctest runner separately?

ezyang commented:
The test command CI is using is

```
python test/run_test.py --exclude-jit-executor --exclude-distributed-tests --shard 1 4 --verbose
```

In the run it is indirectly calling

```
Executing ['/opt/conda/bin/python', '-bb', 'test_ops.py', '-v', '--import-slow-tests', '--import-disabled-tests'] ... [2022-08-12 16:12:50.781112]
```

so it looks like pytest isn't always used. Does xdoctest require pytest? Maybe we can just invoke pytest separately for the doc tests.

Erotemic commented:
No, xdoctest does not require pytest (in fact it runs faster without it). Now that I know where to look, I can probably find a good way of enabling it in run_test.py.

Erotemic committed:
Attempt running doctests

Erotemic commented:
I pushed up something that should at least get them to attempt to run. I'm not sure if it's the best way to do it in the context of `run_tests.py`. It seems like it has it's own grouping and organization of the tests, and right now I just lumped "doctests" as its own special test with its own special handler. This means all of the doctests are run in a single command. 

It looks like run_tests.py might want to be a bit more granular? I could use the xdoctest API to directly parse out each doctest and group them into something logical like this:

```python
        # Parse all doctests out of the entire package
        import pathlib
        from collections import defaultdict
        pkgpath = pathlib.Path(torch.__file__).parent
        doctests = list(xdoctest.core.parse_doctestables(str(pkgpath), style='google'))
        grouped_doctests = defaultdict(list)
        for doctest in doctests:
            relname = pathlib.Path(doctest.fpath).relative_to(pkgpath.parent)
            groupid = str('/'.join(relname.parent.parts[0:2]))
            grouped_doctests[groupid].append(doctest)
        import ubelt as ub
        print(ub.repr2(ub.udict(grouped_doctests).map_values(len)))
```

which results in

```
{
    'torch': 41,
    'torch/ao': 23,
    'torch/autograd': 25,
    'torch/cuda': 4,
    'torch/distributed': 71,
    'torch/distributions': 39,
    'torch/futures': 5,
    'torch/fx': 12,
    'torch/jit': 10,
    'torch/nn': 237,
    'torch/onnx': 1,
    'torch/optim': 19,
    'torch/package': 1,
    'torch/profiler': 1,
    'torch/sparse': 1,
    'torch/testing': 3,
    'torch/utils': 54,
}
```

Where the number is the number of doctests in that group. Of course it doesn't have to be this grouping. But any guidance on how or if the doctests should be split up would be helpful.

Erotemic commented:
Well, it looks like it's enabled now, and it's correctly complaining about some doctest failures. Not sure why I didn't catch those on my local build, but there's only 17 failures, so we should be able to fix that here.

Erotemic committed:
Fixed / improved some doctests, added pygments

vadimkantorov commented:
I guess the next step might be to make tutorials' attached code (at least the final collected code) also test in some way (at least in some smoke test way or e.g. catch that it learns something or loss decreases - maybe some self-check can be incorporated in this tutorial code that checks that loss decreases okay and prints some magic string to the screen which can be tested against)

vadimkantorov commented:
One other thing to test is match of actual argnames versus the argnames in the docs: https://github.com/pytorch/pytorch/issues/54881

ezyang reviewed (approved):
I greatly appreciate all the work you've put into this. Ready to merge whenever CI passes!

Erotemic committed:
Fixes for win32

Erotemic commented:
@vadimkantorov I love tutorials as code. One thing that helps with that is having semi-realistic toy-examples. Towards that end, I want to point you to my work on [kwcoco](https://gitlab.kitware.com/computer-vision/kwcoco) It is an extension of the MS-COCO format, but it includes some other nifty tools. The one I want to point you at is the "toydata". The problem I was solving was: I want to write tests for semi-realistic examples, but they need to work quickly and I can't download a dataset. To that end I wrote a small lightweight system that produces these dummy coco files and can render a semi-interesting looking image to match it.

I'm also interested in adding some sort of variable name checking behavior to xdoctest. In my ["xdoctest future work" slide ](https://docs.google.com/presentation/d/1563XL-n7534QmktrkLSjVqX36z5uhjUFrPw8wIO6z1c/edit#slide=id.g70c0722eb3_2_268) I mention that there are "cool refactoring tools that could be built on top of its framework". I'm currently looking into: 1. Automatically adding the +SKIP directive to failing tests (the prototype of which was developed for this) and 2. Rewrite the got/want whatever-you-call-the expected text doctest strings to be exactly what the runner observed (which would have helped for this, but I did it manually because it doesn't exist yet).

If you're interested in checking if doctests use the same names as the signature of the function, that would have to be built, but xdoctest could aid in extracting the name of the function and the corresponding doctest. Then I imagine some ast could could (often, maybe not 100% due to some undecidable problem) extract the names of the variables used when calling the function, and check that against its signature dynamically. Seems like there would be a lot of edge-cases though. Something similar I was interested in was seeing if I could use the doctests to infer partial type annotations for the function args and return type. Both problems are very similar. This is much lower on my priority list, so if you hear of a solution, let me know.

@ezyang The latest CI seems to be much more green. The last errors seem to be on a win32 machine. Changes:

* added more `dim=1` to prevent warnings.
* skipped a few tests based on `# xdoctest +REQUIRES(POSIX)` and `#xdoctest +REQUIRES(env:TORCH_DOCTEST_ENABLE_QENGINE)` (which means that test will be skipped unless we make that environ). 
* minor formatting

Also I forgot to mention from the last one. My linter flags any class without 2 lines above it. I mostly avoided my temptation to haphazardly lint. I did add a few newlines, but I think it was only in files I was also making other edits in (I can revert those if needed).

Thank you for the recognition. I appreciate being able to contribute my software as a part of this project and helping ensure the examples in the documentation work correctly. Pytorch is a truly extraordinary accomplishment of human engineering. (and extra xdoctest downloads is also a plus! I'll also give a shoutout to my [ubelt](https://github.com/Erotemic/ubelt) project if I'm already blatantly self-promoting).

vadimkantorov commented:
I think the real problem about getting the names of overloads is that PyTorch uses some magical way of generating these Python methods that does dispatch to underlying functions. In that issue @Lezcano seems to have dug this already...

I guess in general some inspect.getargspec can be used, but for PyTorch specifically one needs to check about that dispatch thingies

ezyang commented:
BTW, I'm quite happy to exclude doctest from running on Windows; we only really need to test them on Linux and CUDA

Erotemic committed:
Make all requires based on environs, rework rpc tests

Erotemic commented:
I'll give the win32 tests another shot at working. It seems to only be the RPC and quantized stuff its having trouble with. The xdoctest runner should not even be trying to import the code if it doesn't see any enabled doctests in it, so it might be something I have to patch in xdoctest itself.

Latest commits switches the `REQUIRES(--lapack)` to `REQUIRES(env:TORCH_DOCTEST_LAPACK)`, and I did similar things with other requires tests. I added checks before the doctests run to conditionally enable them. I'm not sure I'm doing the right thing for the quantized checks, so let me know if there is a method to explicitly check for which quantized engines are supported: 

```python   
    # Determine if we should conditionally enable any of these doctests via
    # environment variables.
    if torch.cuda.is_available():
        os.environ['TORCH_DOCTEST_CUDA'] = '1'
        if torch.cuda.device_count() > 1:
            os.environ['TORCH_DOCTEST_CUDA1'] = '1'

    if torch._C.has_lapack:
        os.environ['TORCH_DOCTEST_LAPACK'] = '1'

    try:
        # Is there a better check if quantization is enabled?
        import torch.nn.quantized as nnq  # NOQA
        torch.backends.quantized.engine = 'qnnpack'
        torch.backends.quantized.engine = 'fbgemm'
    except (ImportError, RuntimeError):
        ...
    else:
        os.environ['TORCH_DOCTEST_QENGINE'] = '1'
```

Dashboards should be green in a few more iterations.

Erotemic committed:
check both quantized engines

Erotemic committed:
Fix lapack tests

Erotemic committed:
Disable qengine doctests

Erotemic commented:
So this behavior has happened more than once on `[linux-focal-py3.7-clang7-asan / test (default, 1, 5, linux.2xlarge)](https://github.com/pytorch/pytorch/runs/7891670634?check_suite_focus=true#logs)`. It's possible it's an issue the doctests might have exposed: 

```
* SUCCESS: /opt/conda/lib/python3.7/site-packages/torch/nn/quantized/modules/conv.py::Conv3d:0
* DOCTEST : /opt/conda/lib/python3.7/site-packages/torch/nn/quantized/modules/conv.py::ConvTranspose1d:0, line 681 <- wrt source file
/var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/4x4c2-sse2.c:334:5: runtime error: store to misaligned address 0x6330003b1ac2 for type 'uint32_t' (aka 'unsigned int'), which requires 4 byte alignment
0x6330003b1ac2: note: pointer points here
 d0 3d  7d e8 d8 3d f2 33 ba 3d  a6 55 bc 3d cc 06 ac 3d  31 0c d0 3d 09 3a dc 3d  d8 16 ca 3d a9 cc
              ^ 
    #0 0x7fa36f271b12 in pytorch_q8conv_ukernel_4x4c2__sse2 (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x1c189b12)
    #1 0x7fa36f253a02 in qnnpack::compute_q8conv(qnnpack::q8conv_context const*, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long, unsigned long) (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x1c16ba02)
    #2 0x7fa3737aaede in compute_4d_tiled (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x206c2ede)
    #3 0x7fa36f21ad33 in pthreadpool_thread_parallelize_1d_fastpath (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x1c132d33)
    #4 0x7fa36f219eae in thread_main (/opt/conda/lib/python3.7/site-packages/torch/lib/libtorch_cpu.so+0x1c131eae)
    #5 0x7fa390f3b608 in start_thread (/lib/x86_64-linux-gnu/libpthread.so.0+0x8608)
    #6 0x7fa390e60132 in clone (/lib/x86_64-linux-gnu/libc.so.6+0x11f132)

SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior /var/lib/jenkins/workspace/aten/src/ATen/native/quantized/cpu/qnnpack/src/q8conv/4x4c2-sse2.c:334:5 in 
```

I'm going to try disabling the quantized tests to see if it passes.

I'm also seeing:

```
[0 / 6] [Prepa] .../compiler/xla/xla_client:libxla_computation_client.so

ERROR: /var/lib/jenkins/workspace/xla/third_party/tensorflow/tensorflow/compiler/xla/service/cpu/BUILD:780:11: Compiling tensorflow/compiler/xla/service/cpu/runtime_custom_call_status.cc failed: (Exit 1): clang failed: error executing command /usr/lib/llvm-8/bin/clang -U_FORTIFY_SOURCE -fstack-protector -Wall -Wthread-safety -Wself-assign -fcolor-diagnostics -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections ... (remaining 31 arguments skipped)

Use --sandbox_debug to see verbose messages from the sandbox
In file included from tensorflow/compiler/xla/service/cpu/runtime_custom_call_status.cc:18:
./tensorflow/compiler/xla/service/custom_call_status_internal.h:26:8: error: no template named 'optional' in namespace 'std'
  std::optional<std::string> message;
  ~~~~~^
./tensorflow/compiler/xla/service/custom_call_status_internal.h:34:6: error: no template named 'optional' in namespace 'std'
std::optional<absl::string_view> CustomCallStatusGetMessage(
~~~~~^
2 errors generated.
[3,283 / 11,318] 7 actions, 0 running
    @llvm-project//llvm:DebugInfoCodeView; 0s remote-cache
    @llvm-project//llvm:DebugInfoCodeView; 0s remote-cache
    Compiling jidctint.c; 0s remote-cache
    Compiling linux-x86_64/crypto/test/trampoline-x86_64.S; 0s remote-cache
    @llvm-project//llvm:DebugInfoCodeView; 0s remote-cache
    Compiling src/crypto/x509/x509cset.c; Downloading exte...; 0s remote-cache
    @boringssl//:crypto; 0s remote-cache








Target //tensorflow/compiler/xla/xla_client:libxla_computation_client.so failed to build
[3,290 / 11,318] checking cached actions

Use --verbose_failures to see the command lines of failed build steps.
[3,290 / 11,318] checking cached actions

INFO: Elapsed time: 76.658s, Critical Path: 2.29s
[3,290 / 11,318] checking cached actions

INFO: 3290 processes: 368 remote cache hit, 2922 internal.
[3,290 / 11,318] checking cached actions

FAILED: Build did NOT complete successfully

FAILED: Build did NOT complete successfully
Failed to build external libraries: ['/var/lib/jenkins/workspace/xla/build_torch_xla_libs.sh', '-O', '-D_GLIBCXX_USE_CXX11_ABI=1', 'install']
+ sccache_epilogue
+ echo '::group::Sccache Compilation Log'
```

on `[linux-bionic-py3_7-clang8-xla / test (xla, 1, 1, linux.2xlarge)](https://github.com/pytorch/pytorch/runs/7891669327?check_suite_focus=true#logs)`, but I'm not sure how this patch could be causing that error.

Erotemic committed:
start bisecting

ezyang commented on test/run_test.py:
```diff
@@ -128,7 +128,8 @@ def skip_test_p(name: str) -> bool:
 
-TESTS = TESTS + ['doctests']
+# Q: Is there a better way to add "doctests" to the list of tests?
+# TESTS = TESTS + ['doctests']
```
I guess, conventionally, there has to be a file. `run_test.py` isn't really setup for virtual tests in this way.

Erotemic replied:
If that is the case, and this doesn't map to an exiting pattern, then I think the way I'm doing it here is fine. I added a comment describing why the doctests are a special case.

Erotemic committed:
Tests with all conditional features disabled

ezyang commented:
I think the xla failure is unrelated

Erotemic commented:
It's strange, because the very first run of the CI on this branch was green. 

Either way the windows errors are an xdoctest issue, and something I think I need to fix in xdoctest itself. None of the doctests in the distributed module are enabled, so it should recognize a file as having all skipped doctests and not even try to import it. 

The other errors are due to timeouts and a segmentation fault that occur after the tests finish. That likely means there is a resource that isn't properly cleaned up in one of the doctests.

Erotemic committed:
bump xdoctest to 1.0.2

Erotemic committed:
Fixed more failing doctests

Erotemic committed:
Fix doctest syntax

Erotemic committed:
Fix warnings and syntax errors

Erotemic committed:
Only list and parse doctests for syntax errors by default

Erotemic committed:
remove xdoctest in regular pytest ini

Erotemic commented:
So based on the experiment in #83808 there is something in this PR that is causing the failures. The task now is to identify what that is.

Erotemic committed:
Fix syntax error, and dont fail when only listing doctests

Erotemic commented:
@ezyang The remaining failures caused by enabling the doctests are turning out to be trickier than I anticipated. To get the dashboard green, I've disabled running the doctests, they are still being parsed and checked for syntax errors (all of which I've fixed in this branch).

My thought is that we merge this PR and then work on one more to enable running the doctests for real this time. This will make debugging simpler, and merging the checks as-is will at least warn developers if they add a doctest with invalid Python syntax (the recent ones added in torch/distributed/_shard/checkpoint/planner.py were missing colons). So it will add some benefit, and it will also prevent merge conflicts from happening with the doctests that I fixed in this PR, so the next PR can focus on just getting everything running.

Thoughts? Is there anything in this branch that needs to be addressed before an intermediate merge?

ezyang commented:
I'm happy to do an intermediate merge, although unfortunately it seems one more rebase is needed

Erotemic commented:
rebased

ezyang commented on torch/_tensor_str.py:
```diff
@@ -206,7 +214,7 @@ def _vector_str(self, indent, summarize, formatter1, formatter2=None):
         1, int(math.floor((PRINT_OPTS.linewidth - indent) / (element_length)))
     )
-    char_per_line = element_length * elements_per_line
+    # char_per_line = element_length * elements_per_line  # unused
```
just delete it then lol (don't do it here, can do in follow up)

Erotemic replied:
I get nervous when I see something obvious like this in someone else's code. Is it really unused? Maybe it causes some side effect I'm not aware of? 

In this case I was convinced enough to comment it out, but unsure of myself enough to not delete it outright. But I'll remove it in the follow-up.

ezyang commented on torch/_tensor_str.py:
```diff
@@ -46,12 +46,20 @@ def set_printoptions(
         >>> torch.arange(10)
         tensor([0, 1, 2, ..., 7, 8, 9])
+        >>> # Restore defaults
+        >>> torch.set_printoptions(profile='default')
```
Something that is a bit bothersome is that if the doctest fails, the default restoration here will not occur, and then all subsequent tests will deliver nonsense. A "always run this afterwards no matter what" designation would be a big help.

Erotemic replied:
It is possible to use a try/finally in the doctest itself, but that get's a little ugly. This may be a good usecase for adding an `xdoctest: +finally` directive. But unfortunately there isn't a great way to accomplish doctest cleanup at the moment.

ezyang commented:
@pytorchbot merge
